pynestml/codegeneration/builder.py

```diff
@@ -46,6 +53,54 @@ def __init__(self, target, options: Optional[Mapping[str, Any]]=None):
 
         self._target = target
 
+    def __del__(self):
+        if self.get_option("redirect"):
+            self.get_option("stdout").close()
+            self.get_option("stderr").close()
+
+    def process_output_redirection_(self, options):
+        dict_keys = list(Builder.redirect_build_output.keys())
+        require_redirect_key = dict_keys[0]
+        redirection_path_key = dict_keys[1]
+
+        # default values. The output will be printed to the console.
+        stdout = None
+        stderr = subprocess.STDOUT
+        redirect = False
+        error_location = "stderr"
+
+        if options and len(options) > 0 and require_redirect_key in options and options.get(require_redirect_key, False):
+            output_file_name = f"{self.get_builder_name()}_output.txt"
+            error_file_name = f"{self.get_builder_name()}_error.txt"
+
+            if redirection_path_key in options:
+                if os.path.isdir(options[redirection_path_key]):
+                    stdout = os.path.join(options[redirection_path_key], output_file_name)
+                    stderr = os.path.join(options[redirection_path_key], error_file_name)
+                else:
+                    raise Exception(f"The provided directory {options[redirection_path_key]} does not exist in your system!")
+            else:
+                target_path = FrontendConfiguration.get_target_path()
+                stdout = os.path.join(target_path, output_file_name)
+                stderr = os.path.join(target_path, error_file_name)
+
+            error_location = stderr
+            stdout = open(stdout, "w")
+            stderr = open(stderr, "w")
+            redirect = True
+
+        self.add_options({"stdout": stdout, "stderr": stderr, "redirect": redirect, "error_location": error_location})
+
     @abstractmethod
     def build(self) -> None:
```

Comment:
How about we rename this to ``build_()`` everywhere, and implement a new, non-abstract method ``build()`` in the base class that handles opening and closing the files safely and calls ``build_()`` from within the right context? Something like:
```py
def build():
    stdout, stderr = process_output_redirection_(self.options)
    with open(..., 'w') as stdout, open(..., 'w') as stderr:
        self._build(stdout, stderr)
```
Do you think it makes sense for ``_build()`` to accept the stdout, stderr parameters (in all ``Builder``s)? This means that we would be more or less explicitly requiring all builders to implement this functionality (of redirecting stdout and stderr in their subprocess or other calls).